### PR TITLE
resource/aws_iot_topic_rule: Fix optional attributes

### LINKED ISSUE #6195: Terraform requires arguments that are optional when creating aws_iot_topic_rule
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
Terraform v0.11.8

### Affected Resource(s)

* aws_iot_topic_rule

### Terraform Configuration Files

```hcl
resource "aws_iot_topic_rule" "client_id_mapping" {
  name        = "${var.env_name}_client_id_mapping"
  enabled     = "true"

  sql = <<EOF
SELECT clientId, {"key": "value"} as mapping
FROM '$$aws/events/subscriptions/subscribed/#' 
EOF

  sql_version = "2016-03-23"

  dynamodb {
    hash_key_field  = "client_id"
    hash_key_value  = "$${clientId}"
    hash_key_type   = "STRING"
    range_key_field = ""
    range_key_value = ""
    payload_field   = "$${mapping}"
    table_name      = "${var.env_name}_client_id_mapping"
    role_arn        = "${aws_iam_role.client_id_mapping_role.arn}"
  }
}
```

### Debug Output

```bash
>>> terraform apply --var-file=../../../variables/dev.tfvar
Error: aws_iot_topic_rule.client_id_mapping: "dynamodb.0.range_key_field": required field is not set
Error: aws_iot_topic_rule.client_id_mapping: "dynamodb.0.range_key_value": required field is not set
```

### Expected Behavior

These range_key arguments of the DynamoDB action are optional according to the AWS API reference and the resource's TF documentation.

### Actual Behavior

Both arguments are marked as required in the [source code](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_iot_topic_rule.go).

### References

* [iot_topic_rule TF docs](https://www.terraform.io/docs/providers/aws/r/iot_topic_rule.html)
* [CreateTopicRule API Reference](https://docs.aws.amazon.com/iot/latest/apireference/API_CreateTopicRule.html)
* [DynamoDBAction API Reference]( https://docs.aws.amazon.com/iot/latest/apireference/API_DynamoDBAction.html)

### PR DESCRIPTION
**Pull request by nywilken**
Terraform requires arguments that are optional when creating aws_iot_topic_rule

Logged POST request before update

```                                                                                                                                                         
$ TF_LOG_PATH=log.txt TF_LOG=debug TF_ACC=1 go test ./aws -v -timeout 120m -parallel 10 -run=TestAccAWSIoTTopicRule_basic                                     
                                                                                                                                                              
2019/02/08 12:48:15 [DEBUG] [aws-sdk-go] DEBUG: Request iot/CreateTopicRule Details:                                                                          
---[ REQUEST POST-SIGN ]-----------------------------                                                                                                         POST /rules/test_rule_nlzsp HTTP/1.1
[...]
{"actions":[{"dynamoDB":{"hashKeyField":"hash_key_field","hashKeyValue":"hash_key_value","payloadField":"payload_field","rangeKeyField":"range_key_field","rangeKeyValue":"range_key_value","roleArn":"arn:aws:iam::187416307283:role/test_role_nlzsp","tableName":"table_name"}}],"awsIotSqlVersion":"2015-10-08","description":"Example rule","ruleDisabled":false,"sql":"SELECT * FROM 'topic/test'"} 
-----------------------------------------------------
2019/02/08 12:48:15 [DEBUG] [aws-sdk-go] DEBUG: Response iam/CreatePolicy Details: 
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK^M                                                                                                                                             
```

Logged POST request after update
```
$TF_LOG_PATH=log.txt TF_LOG=debug TF_ACC=1 go test ./aws -v -timeout 120m -parallel 10 -run=TestAccAWSIoTTopicRule_dynamodb 
2019/02/08 12:54:28 [DEBUG] [aws-sdk-go] DEBUG: Request iot/CreateTopicRule Details:
---[ REQUEST POST-SIGN ]-----------------------------
POST /rules/test_rule_6qq07 HTTP/1.1
[...]
{"actions":[{"dynamoDB":{"hashKeyField":"hash_key_field","hashKeyValue":"hash_key_value","payloadField":"payload_field","roleArn":"arn:aws:iam::187416307283:role/test_role_6qq07","tableName":"table_name"}}],"awsIotSqlVersion":"2015-10-08","description":"Example rule","ruleDisabled":false,"sql":"SELECT * FROM 'topic/test'"}
-----------------------------------------------------
2019/02/08 12:54:28 [DEBUG] [aws-sdk-go] DEBUG: Response iam/GetPolicyVersion Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
```

closes #6195

### CODE REVIEW

File: aws/resource_aws_iot_topic_rule.go [hunk 1]
```diff
@@ -124,11 +124,11 @@ func resourceAwsIotTopicRule() *schema.Resource {
 						},
 						"range_key_field": {
 							Type:     schema.TypeString,
-							Required: true,
+							Optional: true,
```
Comment (bflad) [inline aws/resource_aws_iot_topic_rule.go]: Can you please either update any existing acceptance testing to omit these values or preferably add an acceptance test that covers omitting these values successfully? Thanks!

Comment (nywilken) [reply aws/resource_aws_iot_topic_rule.go]: Ah good catch!

**Comment by nywilken** [comment]
@bflad I made a few additional updates along with the added test case. Please let me know if there is anything that I may have missed.

File: aws/resource_aws_iot_topic_rule_test.go [hunk 2]
```diff
@@ -90,6 +90,24 @@ func TestAccAWSIoTTopicRule_cloudwatchmetric(t *testing.T) {
 	})
 }
 
+func TestAccAWSIoTTopicRule_dynamodb(t *testing.T) {
```
Comment (bflad) [inline aws/resource_aws_iot_topic_rule_test.go]: 😍

File: aws/resource_aws_iot_topic_rule_test.go [hunk 3]
```diff
@@ -343,16 +361,6 @@ resource "aws_iot_topic_rule" "rule" {
   sql = "SELECT * FROM 'topic/test'"
   sql_version = "2015-10-08"
 
-  // Fake data
-  dynamodb {
-    hash_key_field = "hash_key_field"
-    hash_key_value = "hash_key_value"
-    payload_field = "payload_field"
-    range_key_field = "range_key_field"
```
Comment (bflad) [inline aws/resource_aws_iot_topic_rule_test.go]: 😅 now we have the inverse problem where `range_key_field`, etc. is not being acceptance tested at all. Could you create a test and test configuration very similar to the new one which includes the newly optional arguments? See also `TestAccAWSIoTTopicRule_firehose_separator` which tests the optional argument separate from `TestAccAWSIoTTopicRule_firehose`. Thanks!

File: aws/structure.go [hunk 4]
```diff
@@ -2834,8 +2834,6 @@ func flattenIoTRuleDynamoDbActions(actions []*iot.Action) []map[string]interface
 		if v != nil {
 			result["hash_key_field"] = *v.HashKeyField
 			result["hash_key_value"] = *v.HashKeyValue
-			result["range_key_field"] = *v.RangeKeyField
```
Comment (bflad) [inline aws/structure.go]: Nit: As a good practice for all these with `*` dereferences around here -- we can save ourselves the trouble of `nil` checking these (and prevent potential panics) using the AWS Go SDK provided helper `aws.StringValue(v.XXX)`

Comment (nywilken) [reply aws/structure.go]: :+1:

**Comment by bflad** [comment]
This has been released in [version 1.59.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1590-february-14-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**